packages/core/src/controllers/verify.ts

```diff
@@ -80,85 +134,108 @@ export class Verify extends IVerify {
-      this.logger.info(`postMessage sent: ${attestationId} ${this.verifyUrl}`);
-    } catch (e) {}
+      this.logger.debug(`fetching public key from: ${this.verifyUrlV2}`);
+      const timeout = this.startAbortTimer(FIVE_SECONDS);
+      const result = await fetch(`${this.verifyUrlV2}/public-key`, {
+        signal: this.abortController.signal,
+      });
+      clearTimeout(timeout);
+      return (await result.json()) as jwk;
+    } catch (e) {
+      this.logger.warn(e);
+    }
+    return undefined;
   };
 
-  private createIframe = async () => {
-    let iframeOnLoadResolve: () => void;
-    const onMessage = (event: MessageEvent) => {
-      if (event.data === "verify_ready") {
-        this.onInit();
-        window.removeEventListener("message", onMessage);
-        iframeOnLoadResolve();
-      }
-    };
-    await Promise.race([
-      new Promise<void>((resolve) => {
-        const existingIframe = document.getElementById(VERIFY_CONTEXT);
-        if (existingIframe) {
-          this.iframe = existingIframe as HTMLIFrameElement;
-          this.onInit();
-          return resolve();
-        }
+  private persistPublicKey = async (publicKey: jwk) => {
+    this.logger.debug(`persisting public key to local storage`, publicKey);
+    await this.store.setItem(this.storeKey, publicKey);
+    this.publicKey = publicKey;
+  };
 
-        window.addEventListener("message", onMessage);
-        const iframe = document.createElement("iframe");
-        iframe.id = VERIFY_CONTEXT;
-        iframe.src = `${this.verifyUrl}/${this.projectId}`;
-        iframe.style.display = "none";
-        document.body.append(iframe);
-        this.iframe = iframe;
-        iframeOnLoadResolve = resolve;
-      }),
-      new Promise((_, reject) =>
-        setTimeout(() => {
-          window.removeEventListener("message", onMessage);
-          reject("verify iframe load timeout");
-        }, toMiliseconds(FIVE_SECONDS)),
-      ),
-    ]);
+  private removePublicKey = async () => {
+    this.logger.debug(`removing verify v2 public key from storage`);
+    await this.store.removeItem(this.storeKey);
+    this.publicKey = undefined;
   };
 
-  private onInit = () => {
-    this.initialized = true;
-    this.processQueue();
+  private isValidJwtAttestation = async (attestation: string) => {
+    const key = await this.getPublicKey();
+    try {
+      if (key) {
+        const validation = await this.validateAttestation(attestation, key);
+        return validation;
+      }
+    } catch (e) {
+      this.logger.warn(e);
+      this.logger.warn("error validating attestation");
+    }
+    const newKey = await this.fetchAndPersistPublicKey();
+    try {
+      if (newKey) {
+        const validation = await this.validateAttestation(attestation, newKey);
+        return validation;
+      }
+    } catch (e) {
+      this.logger.warn(e);
+      this.logger.warn("error validating attestation");
+    }
+    return undefined;
```

Comment:
This could be simplified since `getPublicKey` already calls `fetchAndPersistPublicKey`
```suggestion
    const key = await this.getPublicKey();
    try {
      const validation = await this.validateAttestation(attestation, key);
      return validation;
    } catch (e) {
      this.logger.warn(e);
      this.logger.warn("error validating attestation");
    }
   
    return undefined;
```

Reply:
The current design is intentional in order to refresh the key if a different signing key is provided on the JWT. Perhaps a comment explaining that is needed

Reply:
Hmm I understand the subtle difference now. So if the first case fails we explicitly request a new key.